### PR TITLE
METRON-1845 Correct Test Data Load in Elasticsearch Integration Tests

### PR DESCRIPTION
**Pull request by nickwallen**
The Elasticsearch integration tests use the Elasticsearch API to load the test data before running the tests.  Loading the test data like this does not accurately reflect how the indices will appear in a production environment.

This should be changed to use our existing `ElasticsearchUpdateDao` to write the test data.  This ensures that any changes made to the 'write' portion of our Elasticsearch code will function correctly with the 'read' portion. Said in a different way, this ensures that telemetry written into Elasticsearch by 'Indexing' can be read correctly by the Alerts UI.

- [x] Do not merge until after #1242.  This change is dependent on #1242.  Take a look at the last commit to view the changes specific to this PR.

## Testing

This only changes the integration tests.  If the integration tests pass, we're golden.

## Pull Request Checklist

- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### CODE REVIEW

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java [hunk 1]
```diff
@@ -194,35 +215,41 @@ public Client getClient() {
...
-      Map<String, Object> esDoc = JSONUtils.INSTANCE
-          .load(doc, JSONUtils.MAP_SUPPLIER);
-      indexRequestBuilder.setId((String) esDoc.get(Constants.GUID));
-      Object ts = esDoc.get("timestamp");
-      if (ts != null) {
-        indexRequestBuilder = indexRequestBuilder.setTimestamp(ts.toString());
-      }
-      bulkRequest.add(indexRequestBuilder);
-    }
+  public void add(UpdateDao updateDao, String indexName, String sensorType, Iterable<String> docs)
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: The Elasticsearch tests use this method to load test data.  Rather than using the Transport client here, we use the `UpdateDao` to load the test data.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java [hunk 2]
```diff
@@ -97,48 +118,81 @@ protected static InMemoryComponent startIndex() throws Exception {
...
+
+    // write the test documents for Snort
+    List<String> snortDocuments = new ArrayList<>();
+    for (Object snortObject: (JSONArray) new JSONParser().parse(snortData)) {
+      snortDocuments.add(((JSONObject) snortObject).toJSONString());
     }
-    BulkResponse bulkResponse = bulkRequest.execute().actionGet();
-    if (bulkResponse.hasFailures()) {
-      throw new RuntimeException("Failed to index test data");
+    es.add(updateDao, snortIndex, "snort", snortDocuments);
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: We are now using `ElasticsearchComponent.add` to load the test data here.  This change matches what already exists in the other ES integration tests.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java [hunk 3]
```diff
@@ -97,48 +118,81 @@ protected static InMemoryComponent startIndex() throws Exception {
...
+            .addMapping("snort_doc", JSONUtils.INSTANCE.toJSON(snortTemplate.get("mappings"), false)).get();
+
+    // setup the classes required to write the test data
+    AccessConfig accessConfig = createAccessConfig();
+    ElasticsearchClient client = ElasticsearchUtils.getClient(createGlobalConfig());
+    ElasticsearchRetrieveLatestDao retrieveLatestDao = new ElasticsearchRetrieveLatestDao(client);
+    ElasticsearchColumnMetadataDao columnMetadataDao = new ElasticsearchColumnMetadataDao(client);
+    ElasticsearchRequestSubmitter requestSubmitter = new ElasticsearchRequestSubmitter(client);
+    ElasticsearchUpdateDao updateDao = new ElasticsearchUpdateDao(client, accessConfig, retrieveLatestDao);
+    ElasticsearchSearchDao searchDao = new ElasticsearchSearchDao(client, accessConfig, columnMetadataDao, requestSubmitter);
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: All of this because I need an `UpdateDao` to write the tests data  and a `SearchDao` to make sure the test data is loaded.  Is there a simpler way?

File: metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/UpdateIntegrationTest.java [hunk 4]
```diff
@@ -106,14 +106,14 @@ public void test() throws Exception {
       }}, Optional.empty());
 
       Assert.assertEquals(message0, update.getDocument());
-      Assert.assertEquals(1, getMockHTable().size());
+      Assert.assertEquals(10, getMockHTable().size());
```
Comment (nickwallen) [inline metron-platform/metron-indexing/src/test/java/org/apache/metron/indexing/dao/UpdateIntegrationTest.java]: Since I am reusing the `IndexDao` to load the tests data, the HBase table also gets this test data now.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java [hunk 5]
```diff
@@ -97,45 +118,63 @@ protected static InMemoryComponent startIndex() throws Exception {
...
+    for (Object snortObject: (JSONArray) new JSONParser().parse(snortData)) {
+      snortDocuments.add(((JSONObject) snortObject).toJSONString());
     }
-    BulkResponse bulkResponse = bulkRequest.execute().actionGet();
-    if (bulkResponse.hasFailures()) {
-      throw new RuntimeException("Failed to index test data");
+    es.add(updateDao, snortIndex, "snort", snortDocuments);
+
+    // wait until the test documents are visible
+    assertEventually(() -> Assert.assertEquals(10, findAll(searchDao).getTotal()));
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: Unless I added this to hold up the test and wait for the tests data to be 'visible', some of the tests would reliably fail because the test data wasn't loaded.

**Comment by nickwallen** [comment]
I have merged this with master to pull in the changes from #1242.  This is ready for review.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java (hunk 5 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: @nickwallen is it possible to parameterize the call to do something similar to this? https://github.com/apache/metron/pull/1242/files#diff-99b67fd77500d6232462dc7c27ecff67R148

It's possible this is only available on batch calls, but I wanted to mention it in case we can get away with the assertEventually calls.

Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: I have tried to do exactly this on a private branch that is closer to the code base in #1269.  I made the integration tests supply a RefreshPolicy that was supposed to hold them up until the loaded data is discoverable/searchable.  I was never able to get that to work consistently though despite what the ES docs say.  I still needed the `assertEventually` for the integration tests to pass consistently.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java (hunk 3 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: Oh wow, do we not have a master factory for stitching these all together in the desired default manner? Not sure if this helps, but how bout this -> https://github.com/apache/metron/blob/master/metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchDao.java#L97

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java (hunk 1 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: Might it be better to just use IndexDao, which `extends UpdateDao, SearchDao, RetrieveLatestDao, ColumnMetadataDao`? To that end, if we're looking to route all of this through the ES component in that fashion, it might make sense to simply replace the internal `private Client client;` and instead use the new desired IndexDao for the proxied calls to ES. It could be setup at construction time of the component and remove the need to do the same thing for every test that uses the component class, unless they actually want to do something custom and pass in their own dao during init.

File: metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java [hunk 6]
```diff
@@ -186,4 +195,114 @@ public void testHugeErrorFields() throws Exception {
...
+        Assert.assertEquals(2, columns.size());
+        Assert.assertEquals(message0, JSONUtils.INSTANCE.load(new String(columns.lastEntry().getValue())
+                , JSONUtils.MAP_SUPPLIER)
+        );
+        Assert.assertNotEquals(message0, JSONUtils.INSTANCE.load(new String(columns.firstEntry().getValue())
+                , JSONUtils.MAP_SUPPLIER)
+        );
+      }
+      {
+        //ensure ES is up-to-date
```
Comment (mmiklavc) [inline metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java]: Solr

File: metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java [hunk 7]
```diff
@@ -186,4 +195,114 @@ public void testHugeErrorFields() throws Exception {
     exception.expectMessage("Document contains at least one immense term in field=\"error_hash\"");
     getDao().update(errorDoc, Optional.of("error"));
   }
+
+  @Test
+  @Override
+  public void test() throws Exception {
```
Comment (mmiklavc) [inline metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java]: What do these changes have to do with ES DAO read/write approach change? This class is testing Solr - should this be in a separate PR?

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java (hunk 5 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: Are you able to get this to fail consistently? I ran it a number of times and couldn't get it to fail. I'm wondering if there are any other interesting entries in the logs when this occurs. These tests from ES might be of help - https://github.com/elastic/elasticsearch/pull/17986/files#diff-1c9f982dbd2f9ddb2853d135884621b5R112

Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: > Are you able to get this to fail consistently? I ran it a number of times and couldn't get it to fail.

Can you describe what you did? You just removed line 190 and the tests continue to work reliably?

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java (hunk 1 above)
Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: > Might it be better to just use IndexDao, which extends UpdateDao, SearchDao, RetrieveLatestDao, ColumnMetadataDao? 

Why would that be better?  It needs to do updates, so it needs an `UpdateDao`.  Seemed logical to me.

File: metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java (hunk 7 above)
Comment (nickwallen) [reply metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java]: This test was previously shared in UpdateIntegrationTest between ES and Solr.  With these changes, the tests don't behave exactly the same anymore.  That being said, what I did here doesn't look right.   I'll dig into this and see what is going on.

Comment (nickwallen) [reply metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java]: Ok, I found the problem. With my changes, the test data was getting loaded into both the ElasticsearchDao and the HBaseDao because I was using a MultiIndexDao to load the test data.  

In master, only Elasticsearch gets loaded with the test data.  This was causing some of the test assumptions to fail.  I corrected how the data is loaded so it only loads Elasticsearch and matches the existing behavior.

Thanks for pointing this out.

File: metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java [hunk 8]
```diff
@@ -186,4 +195,114 @@ public void testHugeErrorFields() throws Exception {
...
+        Result r = getMockHTable().get(g);
+        NavigableMap<byte[], byte[]> columns = r.getFamilyMap(CF.getBytes());
+        Assert.assertEquals(1, columns.size());
+        Assert.assertEquals(message0
+                , JSONUtils.INSTANCE.load(new String(columns.lastEntry().getValue())
+                        , JSONUtils.MAP_SUPPLIER)
+        );
+      }
+      {
+        //ensure ES is up-to-date
```
[the remark this answers is not in the document]
Comment (nickwallen) [reply metron-platform/metron-solr/src/test/java/org/apache/metron/solr/integration/SolrUpdateIntegrationTest.java]: No longer a problem.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java (hunk 1 above)
Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: > To that end, if we're looking to route all of this through the ES component in that fashion, it might make sense to simply replace the internal private Client client; and instead use the new desired IndexDao for the proxied calls to ES.

I don't think we're ready to do all that quite yet.  There is still some legacy functionality in `ElasticSearchComponent` that uses the underlying client for the old Admin API.  See [close](https://github.com/apache/metron/blob/fcd644ca77394d48d460c460b672a23d6594f49b/metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java#L292), [createIndexWithMapping](https://github.com/apache/metron/blob/fcd644ca77394d48d460c460b672a23d6594f49b/metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java#L228), and [start](https://github.com/apache/metron/blob/fcd644ca77394d48d460c460b672a23d6594f49b/metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java#L153).

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java (hunk 5 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/ElasticsearchSearchIntegrationTest.java]: I meant on the es client change PR, not on this PR. Yes, I had removed assertEventually calls and it consistently passed for me in the search integration test.

File: metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java (hunk 1 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/test/java/org/apache/metron/elasticsearch/integration/components/ElasticSearchComponent.java]: The IndexDao handles all of the plumbing for the init that you're duplicating in this test class.

**Comment by mmiklavc** [comment]
As a side note, I tend to like the approach of using the DAO layers to read/write the way you've done here @nickwallen . I've used this approach in the past, and it helped with test coverage and simplifying the amount of extra custom test code required. It might be worth an overall discussion on how we want to architect testing other similar endpoints going forward.

**Comment by mmiklavc** [comment]
lgtm @nickwallen, +1 pending Travis

**Comment by mmiklavc** [comment]
Still +1 on latest commit - that's a nice improvement. Glad the retry policy ended up working out!
